File: aws-globalaccelerator-listener/src/test/java/software/amazon/globalaccelerator/listener/ReadHandlerTest.java
package software.amazon.globalaccelerator.listener;

import com.amazonaws.services.globalaccelerator.model.ClientAffinity;
import com.amazonaws.services.globalaccelerator.model.DescribeListenerRequest;
import com.amazonaws.services.globalaccelerator.model.DescribeListenerResult;
import com.amazonaws.services.globalaccelerator.model.Listener;
import com.amazonaws.services.globalaccelerator.model.ListenerNotFoundException;
import com.amazonaws.services.globalaccelerator.model.Protocol;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.HandlerErrorCode;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.OperationStatus;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;

import java.util.ArrayList;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class ReadHandlerTest {

    @Mock
    private AmazonWebServicesClientProxy proxy;

    @Mock
    private Logger logger;

    @BeforeEach
    public void setup() {
        proxy = mock(AmazonWebServicesClientProxy.class);
        logger = mock(Logger.class);
    }

    @Test
    public void handleRequest_Read_ReturnsCorrectlyMappedModel() {
        // All we really care about is what describe returns to confirm the mapping is
        // working as expected
        val describeListenerResult = new DescribeListenerResult()
                .withListener(new Listener()
                        .withListenerArn("LISTENER_ARN")
                        .withProtocol(Protocol.TCP.toString())
                        .withClientAffinity(ClientAffinity.SOURCE_IP)
                        .withPortRanges(new com.amazonaws.services.globalaccelerator.model.PortRange().withFromPort(90).withToPort(90)));
        doReturn(describeListenerResult).when(proxy).injectCredentialsAndInvoke(any(DescribeListenerRequest.class), any());

        val portRanges = new ArrayList<PortRange>();
        portRanges.add(new software.amazon.globalaccelerator.listener.PortRange(90, 91));
        val model = ResourceModel.builder()
                .listenerArn("LISTENER_ARN")
                .acceleratorArn("ACCELERATOR_ARN")
                .portRanges(portRanges)
                .build();

        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val response = new ReadHandler().handleRequest(proxy, request, null, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.SUCCESS);
        assertThat(response.getCallbackContext()).isNull();
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();

        // want to confirm data was updated correctly
        assertThat(response.getResourceModel()).isNotNull();
        assertThat(response.getResourceModel().getAcceleratorArn()).isEqualTo("ACCELERATOR_ARN");
        assertThat(response.getResourceModel().getProtocol()).isEqualTo(Protocol.TCP.toString());
        assertThat(response.getResourceModel().getClientAffinity()).isEqualTo(ClientAffinity.SOURCE_IP.toString());
    }

    @Test
    public void handleRequest_MissingListener_ReturnsFailureAndNullModel() {
        doThrow(new ListenerNotFoundException("NOT FOUND")).when(proxy).injectCredentialsAndInvoke(any(DescribeListenerRequest.class), any());

        val portRanges = new ArrayList<PortRange>();
        portRanges.add(new software.amazon.globalaccelerator.listener.PortRange(90, 91));
        val model = ResourceModel.builder()
                .listenerArn("LISTENER_ARN")
                .acceleratorArn("ACCELERATOR_ARN")
                .portRanges(portRanges)
                .build();

        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val response = new ReadHandler().handleRequest(proxy, request, null, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.FAILED);
        assertThat(response.getErrorCode()).isEqualTo(HandlerErrorCode.NotFound);
        assertThat(response.getResourceModel()).isNull();
    }
}

File: aws-globalaccelerator-listener/src/test/kotlin/software/amazon/globalaccelerator/listener/ReadHandlerTest.kt
package software.amazon.globalaccelerator.listener

import com.amazonaws.services.globalaccelerator.model.ClientAffinity
import com.amazonaws.services.globalaccelerator.model.DescribeListenerRequest
import com.amazonaws.services.globalaccelerator.model.DescribeListenerResult
import com.amazonaws.services.globalaccelerator.model.Listener
import com.amazonaws.services.globalaccelerator.model.ListenerNotFoundException
import com.amazonaws.services.globalaccelerator.model.Protocol
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.OperationStatus
import software.amazon.cloudformation.proxy.ResourceHandlerRequest
import java.util.ArrayList


@ExtendWith(MockKExtension::class)
class ReadHandlerTest {
    @MockK
    lateinit var proxy: AmazonWebServicesClientProxy

    @MockK(relaxed = true)
    lateinit var logger: Logger

    @BeforeEach
    fun setup() = MockKAnnotations.init(this)

    private val acceleratorArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d"
    private val listenerArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d/listener/ee7358c2"

    @Test
    fun handleRequest_Read_ReturnsCorrectlyMappedModel() {
        val describeListenerResult = DescribeListenerResult()
                .withListener(Listener()
                        .withListenerArn(listenerArn)
                        .withProtocol(Protocol.TCP.toString())
                        .withClientAffinity(ClientAffinity.SOURCE_IP)
                        .withPortRanges(com.amazonaws.services.globalaccelerator.model.PortRange().withFromPort(90).withToPort(90)))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeListener>()) } returns describeListenerResult

        val portRanges = ArrayList<PortRange>()
        portRanges.add(PortRange(90, 91))
        val model = ResourceModel.builder()
                .listenerArn(listenerArn)
                .acceleratorArn(acceleratorArn)
                .portRanges(portRanges)
                .build()

        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val response = ReadHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.SUCCESS, response.status)
        assertNull(response.callbackContext)
        assertNull(response.resourceModels)
        assertNull(response.message)

        assertNotNull(response.resourceModel)
        assertEquals(acceleratorArn, response.resourceModel.acceleratorArn)
        assertEquals(Protocol.TCP.toString(), response.resourceModel.protocol)
        assertEquals(ClientAffinity.SOURCE_IP.toString(), response.resourceModel.clientAffinity)

    }

    @Test
    fun handleRequest_MissingListener_ReturnsFailureAndNullModel() {
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeListener>()) } throws(ListenerNotFoundException("NOT FOUND"))

        val portRanges = ArrayList<PortRange>()
        portRanges.add(PortRange(90, 91))
        val model = ResourceModel.builder()
                .listenerArn(listenerArn)
                .acceleratorArn(acceleratorArn)
                .portRanges(portRanges)
                .build()

        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val response = ReadHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.FAILED, response.status)
        assertEquals(HandlerErrorCode.NotFound, response.errorCode)
        assertNull(response.resourceModel)
    }
}

